File: app/src/main/java/com/iita/akilimo/views/fragments/dialog/OperationTypeDialogFragment.java
package com.iita.akilimo.views.fragments.dialog;

import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.RadioGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.iita.akilimo.R;
import com.iita.akilimo.inherit.BaseDialogFragment;
import com.iita.akilimo.interfaces.IDismissOperationsDialogListener;
import com.iita.akilimo.utils.enums.EnumOperationType;


public class OperationTypeDialogFragment extends BaseDialogFragment {
    public static final String ARG_ITEM_ID = "OperationTypeDialogFragment";
    public static final String OPERATION_TYPE = "operation_type";
    private static final String LOG_TAG = OperationTypeDialogFragment.class.getSimpleName();


    private Button btnClose;
    private Button btnUpdate;
    private Button btnRemove;
    private RadioGroup radioGroup;
    private TextView lblSelectionError;

    private Dialog dialog;
    private IDismissOperationsDialogListener onDismissListener;
    private EnumOperationType enumOperationType;
    private String operation;
    private boolean cancelled;

    public OperationTypeDialogFragment(Context context) {
        this.context = context;
    }


    @NonNull
    @Override
    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {

        Bundle bundle = this.getArguments();
        if (bundle != null) {
            operation = bundle.getString(OPERATION_TYPE);
        }
        dialog = new Dialog(context);

        dialog.getWindow().requestFeature(Window.FEATURE_NO_TITLE);
        dialog.getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_LAYOUT_IN_SCREEN);
        dialog.setContentView(R.layout.fragment_operation_type_dialog);

        dialog.setCancelable(false);
        dialog.setCanceledOnTouchOutside(false);
        dialog.getWindow().setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
        dialog.getWindow().getAttributes().windowAnimations = R.style.DialogSlideAnimation;

        btnClose = dialog.findViewById(R.id.close_button);
        btnUpdate = dialog.findViewById(R.id.update_button);
        btnRemove = dialog.findViewById(R.id.remove_button);
        radioGroup = dialog.findViewById(R.id.rdgOperationType);
        lblSelectionError = dialog.findViewById(R.id.lblError);

        TextView lblFragmentTitle = dialog.findViewById(R.id.lblFragmentTitle);
        if (operation.equals("Plough")) {
            lblFragmentTitle.setText(R.string.lbl_plough_op_type);
        } else if (operation.equals("Ridge")) {
            lblFragmentTitle.setText(R.string.lbl_ridge_op_type);
        }

        btnUpdate.setOnClickListener(view -> {
            if (enumOperationType != null) {
                cancelled = false;
                dismiss();
            }
            lblSelectionError.setVisibility(View.VISIBLE);
        });

        btnClose.setOnClickListener(view -> {
            cancelled = true;
            enumOperationType = EnumOperationType.NONE;
            dismiss();
        });

        radioGroup.setOnCheckedChangeListener((radioGroup, radioIndex) -> radioSelected(radioIndex));
        return dialog;
    }

    private void radioSelected(int radioIndex) {
        lblSelectionError.setVisibility(View.GONE);
        switch (radioIndex) {
            case R.id.rdMechanical:
                enumOperationType = EnumOperationType.MECHANICAL;
                break;
            case R.id.rdManual:
                enumOperationType = EnumOperationType.MANUAL;
                break;
            default:
                enumOperationType = EnumOperationType.NONE;
                break;
        }
    }

    @Override
    public void onDismiss(@NonNull DialogInterface dialog) {
        super.onDismiss(dialog);
        if (onDismissListener != null && enumOperationType != null) {
            onDismissListener.onDismiss(operation, enumOperationType, cancelled);
        }
    }

    public void setOnDismissListener(IDismissOperationsDialogListener dismissListener) {
        this.onDismissListener = dismissListener;
    }
}

File: app/src/main/java/com/akilimo/mobile/views/fragments/dialog/OperationTypeDialogFragment.kt
package com.akilimo.mobile.views.fragments.dialog

import android.app.Dialog
import android.content.DialogInterface
import android.graphics.Color
import android.os.Bundle
import android.view.View
import android.view.Window
import androidx.core.graphics.drawable.toDrawable
import com.akilimo.mobile.R
import com.akilimo.mobile.databinding.FragmentOperationTypeDialogBinding
import com.akilimo.mobile.inherit.BaseDialogFragment
import com.akilimo.mobile.interfaces.IDismissOperationsDialogListener
import com.akilimo.mobile.utils.enums.EnumOperation
import com.akilimo.mobile.utils.enums.EnumOperationMethod

class OperationTypeDialogFragment : BaseDialogFragment() {

    private var onDismissListener: IDismissOperationsDialogListener? = null
    private var selectedOperationType: EnumOperationMethod = EnumOperationMethod.NONE
    private var operationType: EnumOperation = EnumOperation.NONE
    private var wasCancelled = false

    private var _binding: FragmentOperationTypeDialogBinding? = null
    private val binding get() = _binding!!

    companion object {
        const val TAG_OPERATION_DIALOG = "OperationTypeDialogFragment"
        const val OPERATION_TYPE = "operation_type_enum"
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val context = requireContext()
        operationType = arguments?.getParcelable(OPERATION_TYPE) ?: EnumOperation.NONE

        _binding = FragmentOperationTypeDialogBinding.inflate(layoutInflater)

        return Dialog(context).apply {
            requestWindowFeature(Window.FEATURE_NO_TITLE)
            window?.apply {
                setBackgroundDrawable(Color.TRANSPARENT.toDrawable())
                attributes.windowAnimations = R.style.DialogSlideAnimation
            }

            setContentView(binding.root)
            setCancelable(true)
            setCanceledOnTouchOutside(false)
            setupUI()
        }
    }

    private fun setupUI() {
        val titleRes = when (operationType) {
            EnumOperation.TILLAGE -> R.string.lbl_plough_op_type
            EnumOperation.RIDGING -> R.string.lbl_ridge_op_type
            else -> R.string.app_name
        }
        binding.lblFragmentTitle.setText(titleRes)

        binding.updateButton.setOnClickListener {
            if (selectedOperationType != EnumOperationMethod.NONE) {
                wasCancelled = false
                dismiss()
            } else {
                binding.lblError.visibility = View.VISIBLE
            }
        }

        binding.closeButton.setOnClickListener {
            wasCancelled = true
            selectedOperationType = EnumOperationMethod.NONE
            dismiss()
        }

        binding.rdgOperationType.setOnCheckedChangeListener { _, checkedId ->
            binding.lblError.visibility = View.GONE
            selectedOperationType = when (checkedId) {
                R.id.rdMechanical -> EnumOperationMethod.TRACTOR
                R.id.rdManual -> EnumOperationMethod.MANUAL
                else -> EnumOperationMethod.NONE
            }
        }
    }

    override fun onDismiss(dialog: DialogInterface) {
        super.onDismiss(dialog)
        onDismissListener?.onDismiss(selectedOperationType, wasCancelled)
    }

    fun setOnDismissListener(listener: IDismissOperationsDialogListener?) {
        onDismissListener = listener
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}
